drafts/two-way-binding-signals-and-query-params.md

```diff
@@ -0,0 +1,754 @@
+
+We could argue that we are indicating Angular to `merge` the query params in that `router.navigate` call. So no query params should be removed from the browser url.
+
+But again, Angular only emits a new value in the `activatedRoute.queryParams` observable as soon as the current navigation is completed. During start up, the query params object that Angular maintains is just an empty one `{}`. And, when the effect replaces the initial navigation with an "empty navigation", ultimately we will be indicating Angular to update the current query params object from `{}` with `{}`, resulting in, of course `{}`. After the navigation is completed, Angular just reflects the state of the query params in the browser url, and since it's `{}`, all query params are removed from the browser url. 
+
+Since there's a lot going on, let's try to visualize this cancelling navigation idea. We're going to add some debug code and simplify the usage in the `AppComponent`:
+
+```ts
+@Injectable({ providedIn: "root" })
+export class QueryParamService {
+  private activatedRoute = inject(ActivatedRoute);
+  private router = inject(Router);
+
+  private allQueryParams = toSignal(this.activatedRoute.queryParams);
+
+  constructor() {
+    // Notice that we log navigation events here
+    this.router.events.subscribe((e) => console.log(e));
+  }
+
+  // Same as before
+  get(queryParamName: string) {
+    const activatedRouteValue = computed(() => this.allQueryParams()?.[queryParamName]);
+    const queryParamValue = signal(activatedRouteValue());
+
+    effect(() => queryParamValue.set(activatedRouteValue()), { allowSignalWrites: true });
+
+    effect(() =>
+      this.router.navigate([], {
+        queryParams: { [queryParamName]: queryParamValue() },
+        queryParamsHandling: "merge",
+      })
+    );
+
+    return queryParamValue;
+  }
+}
+```
+
+```ts
+@Component({
+  selector: "app-root",
+  standalone: true,
+  template: `
+    <h1>Signals Demo</h1>
+    <p>Your first name is: {{ firstName() }}</p>
+  `,
+})
+export class AppComponent {
+  private queryParamService = inject(QueryParamService);
+
+  // Notice we only use the .get() method once
+  firstName = this.queryParamService.get("firstName");
+}
+```
+
+![Angular cancelling navigations](https://cdn.hashnode.com/res/hashnode/image/upload/v1710594428122/YXorJBf0j.gif?auto=format)
+
+Notice that when we visit the url in the browser, Angular triggers an initial navigation on start up, this initial navigation contains the query params actual values, those that are initially provided in the browser url. But, as soon as we execute the second effect inside the `queryParamService.get` method, a new empty navigation is triggered, which cancels the initial one, losing with it all query params data.
```

Comment:
**thought:** again, this is just digging into the same problem for the 3rd(?) time, let's go to the solution quicker

I could see this be a separate article to be honest, something like "Why initial Angular Query Params are missing" or "How Angular Active Route works" or something like that. It's an interesting topic, but makes this article lose focus.